### PR TITLE
Handle workspaces in computeCodePaths

### LINKED ISSUE #15436: Handle yarn/npm workspaces when computing code paths for function serialisation
When we compute the code paths for function serialisation, we are not handling yarn/npm workspace setups where the node_modules might be in a different directory.

Related issues:
* https://github.com/pulumi/pulumi/issues/7168
* https://github.com/pulumi/pulumi/issues/2661
* https://github.com/pulumi/pulumi/issues/3013

### PR DESCRIPTION
**Pull request by julienp**
# Description

When we compute the code paths for function serialisation, we are not handling yarn/npm workspace setups where the node_modules might be in a different directory.

Fixes #15436
Fixes https://github.com/pulumi/pulumi/issues/2661
Fixes https://github.com/pulumi/pulumi/issues/7168
Fixes https://github.com/pulumi/pulumi/issues/3013

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/testing/integration/program.go [hunk 1]
```diff
@@ -2171,6 +2182,86 @@ func (pt *ProgramTester) prepareNodeJSProject(projinfo *engine.Projinfo) error {
 	return nil
 }
 
+var errNotInWorkspace = errors.New("not in a workspace")
+
+func findWorkspaceRoot(programDirectory string) (string, error) {
```
Comment (julienp) [inline pkg/testing/integration/program.go:2187]: TODO: Copied from here https://github.com/pulumi/pulumi/pull/15421/files#diff-9f95bbceb9df8458a048c747ac37385bd37d97c012c640456cad2604aeed85dbR835

File: sdk/nodejs/runtime/closure/codePaths.ts [hunk 2]
```diff
@@ -179,78 +193,118 @@ function searchUp(currentDir: string, fileToFind: string): string | null {
...
-            throw new ResourceError("Failed to find package.json.", logResource);
+    // the working directory is the directory containing the package.json file
+    let workingDir = searchUp(".", "package.json");
+    if (workingDir === null) {
+        // we couldn't find a directory containing package.json
+        // searching up from the current directory
+        throw new ResourceError("Failed to find package.json.", logResource);
+    }
+    // Ensure workingDir is a relative path so we get relative paths in the output.
+    workingDir = upath.relative(upath.resolve("."), workingDir);
```
Comment (julienp) [inline sdk/nodejs/runtime/closure/codePaths.ts:250]: This was not the case previously, and the uploaded lambda ended up dropping the dependencies.

File: sdk/nodejs/npm/workspaces.go [hunk 3]
```diff
@@ -48,11 +48,11 @@ func FindWorkspaceRoot(programDirectory string) (string, error) {
 		}
 		for _, workspace := range workspaces {
 			// See if any of the workspace glob results is the programDirectory.
-			paths, err := filepath.Glob(filepath.Join(currentDir, workspace))
+			paths, err := filepath.Glob(filepath.Join(currentDir, workspace, "package.json"))
```
Comment (julienp) [inline sdk/nodejs/npm/workspaces.go:51]: By including `package.json` in the pattern we limit the number of paths returned to those we're really interested in.

**Review by Frassle — COMMENTED** [review]
Feels like it would be worth having a test with a dynamic provider to ensure code execution of serialized closures actually works in this workspace setups.

**Comment by julienp** [comment]
> Feels like it would be worth having a test with a dynamic provider to ensure code execution of serialized closures actually works in this workspace setups.

It looks like dynamic provider does not call computeCodePaths at all, only serializeFunction. I am looking through the code and it is *interesting*. I guess that the dynamic provider process that executes the serialised function is being started in the same directory as the pulumi program and that's what we rely on for having the right node_modules in place.

**Review by Frassle — APPROVED** [review]

**Pull request merged**